programs/pluto/kernel_xfrm.c

```diff
@@ -1906,6 +1905,70 @@ static void process_addr_chage(struct nlmsghdr *n, struct logger *logger)
 		rta = RTA_NEXT(rta, msg_size);
 	}
 }
+static void netlink_kernel_sa_expire(struct nlmsghdr *n, struct logger *logger)
+{
+	struct xfrm_user_expire *ue = NLMSG_DATA(n);
+
+	if (n->nlmsg_len < NLMSG_LENGTH(sizeof(*ue))) {
+		llog(RC_LOG, logger,
+			"netlink_expire got message with length %zu < %zu bytes; ignore message",
+			(size_t) n->nlmsg_len, sizeof(*ue));
+		return;
+	}
+
+	const struct ip_info *afi = aftoinfo(ue->state.family);
+	if (afi == NULL) {
+		llog(RC_LOG, logger,
+		     "kernel: XFRM_MSG_EXPIRE message malformed: family %u unknown",
+		     ue->state.family);
+		return;
+	}
+
+	const ip_protocol *protocol = protocol_by_ipproto(ue->state.id.proto);
+	if (protocol == NULL) {
+		llog(RC_LOG, logger,
+		      "XFRM_MSG_EXPIRE message from kernel malformed: protocol %u unknown",
+		       ue->state.id.proto);
+		return;
+	}
+
+	ip_address src, dst;
+	address_buf a;
+	address_buf b;
+	xfrm2ip(&ue->state.saddr, &src, ue->state.family);
+	xfrm2ip(&ue->state.id.daddr, &dst, ue->state.family);
+	// dbg("%s spi 0x%x src %s dst %s %s mode %u proto %d bytes %" PRIu64 "packets %" PRIu64,
+	//  error: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 9 has type ‘__u64’ {aka ‘long long unsigned int’} [-Werror=format=]
+	//  bytes and packets complain about %PRIu64 FIX llu to work on 32bit ???
+	dbg("%s spi 0x%x src %s dst %s %s mode %u proto %d bytes %llu packets %llu%s",
+	    __func__, ntohl(ue->state.id.spi),
+	    str_address(&src, &a), str_address(&dst, &b), ue->hard ? "hard" : "soft",
+	    ue->state.mode, ue->state.id.proto, ue->state.curlft.bytes,
+	    ue->state.curlft.packets,
+	    ue->state.id.spi == 0 ? " ACQUIRE state expired discard this message" : "")
+
+	uint8_t protoid = PROTO_RESERVED;
+	switch (ue->state.id.proto) {
+	case  IPPROTO_ESP: protoid = PROTO_IPSEC_ESP; break;
+	case  IPPROTO_AH: protoid = PROTO_IPSEC_AH; break;
+	default:
+		bad_case(ue->state.id.proto);
+	}
+
+	if ((ue->hard && impair.ignore_hard_expire) ||
+	    (!ue->hard && impair.ignore_soft_expire)) {
```

Comment:
this "hard" should be "soft"

Reply:
which one? the second one (!ue->hard)? ue->hard is a kernel variable. there is no soft. there

!ue->hard == soft!